src/python/pants/engine/target.py

```diff
@@ -48,39 +49,62 @@
 # -----------------------------------------------------------------------------------------------
 
 # Type alias to express the intent that the type should be immutable and hashable. There's nothing
-# to actually enforce this, outside of convention. Maybe we could develop a MyPy plugin?
-ImmutableValue = Any
+# to enforce immutability, outside of convention. Maybe we could develop a MyPy plugin?
+ImmutableValue = Hashable
```

Comment:
I thought I had issues with `Hashable`, that it caused a ton of MyPy issues. If this works fine, then great. Otherwise, I recommend going back to `Any`.

If we keep `Hashable`, I think we can remove the type alias and directly use `Hashable` everywhere. One less layer of indirection.

Reply:
That makes sense! My concern would be that it might be more annoying to update this later if we wanted to develop a mypy plugin for immutability (which I wanted to propose as a followup to this). I suppose nobody is actually importing this though so it's probably local to the file? And I think the point of this PR is to allow the `Field` base class to validate the types via its definition -- so `Hashable` seems very reasonable here.